blog/content/posts/client_registration.md

```diff
@@ -0,0 +1,165 @@
+
+This provides the servers with a way to trust the client by deciding which JWKS to trust (or more likely the domain that hosts the JWKS).  It also allows the server to issue the same client ID for the same client used in different places, since it can trust the request is from the same client application. This addresses all the server issues above.
+
+The cost for this implementation is:
+* The client must host a JWKS on a https URL somewhere that the server will trust (even if the client isn't on the web e.g. a native application)
+* The client must craft and sign a software statement
+  * If the client is a native application, this requires using client specific authentication to gate access to the signing key on a remote endpoint.
+* The server must fetch the JWKS URI and verify the statement during registration.
+
+
+This is discussed more in depth in this Specification Enhancement Proposal (SEP) (TODO: link to SEP)
+
+### Client ID Metadata Documents
+
+Another promising solution to this problem is called "Client ID Metadata Documents" (described in [this draft RFC](https://datatracker.ietf.org/doc/draft-parecki-oauth-client-id-metadata-document/) and implemented by Bluesky).
+
+In this approach, we skip the registration step altogether, and provide an https metadata URL as the client ID directly.  The server then fetches the metadata from the URL and uses it as the client's metadata. 
+
+```mermaid
+sequenceDiagram
+   participant Client
+   participant User
+   participant AuthServer
+   participant MetadataURL
+
+   Client->>User: Start OAuth flow
+   User->>AuthServer: Authorization request<br/>(client_id=https://app.com/oauth.json)
+   AuthServer->>MetadataURL: GET https://app.com/oauth.json
+   MetadataURL-->>AuthServer: {name: "App", redirect_uris: [...]}
+   AuthServer->>User: Show consent screen<br/>("App wants access...")
+   User->>AuthServer: Approve
+   AuthServer-->>Client: Authorization code
+   Client->>AuthServer: Exchange code for token<br/>(client_id=https://app.com/oauth.json)
+   AuthServer-->>Client: Access token
+```
+
+Checking against our goals:
+* The servers can trust the client metadata by trusting the domain the metadata is hosted on
+* Servers will have a single client ID per client (the metadata URL)
+* Servers can allow or deny clients based on their metadata URL or metadata domain
+* Servers don't need to handle any database by default, as they can fetch the metadata at authorization time.
+
+The costs for this implementation are:
+* Clients need to host a metadata document on an https URL
+* Clients must provide the URL as their client ID (if they determine an authorization server supports this)
+* Servers need to fetch the metadata, opening egress on their authorization request
+
+Client ID metadata can also be extended to require a JWKS and signed attestation at request time if desired.
```

Comment:
This flow assumes the client can authenticate to the authorization server, typically using keys referenced in the metadata. For cross-domain scenarios (like desktop apps connecting to third-party servers), clients often cannot securely possess the metadata domain's private keys, creating a fundamental constraint on when this pattern works securely. Probably worth calling out.